Add public registration

DanielBiegler opened the pull request:
## Scope

What's changed:

- New UI so you can go to register
- It adds a new user with a specified role
- Fixed a UI bug on the login page
- DB migration that adds fields to directus_settings
- Hide register button depending on directus_settings
- Email validation via customizable filter
- Add mailing service to send out the verification code
- Email template for the verification email
- New per-ip-rate-limiter for registration, their env vars and docs
- Remove `RATE_LIMITER_GLOBAL_STORE` both from code and docs since we dont use it. Lets stick to one store for all limiters.

| Settings | Email |
|--------|--------|
| ![image](https://github.com/directus/directus/assets/14810858/28e9efef-c9b6-424f-99c9-6f59398a9cf8) | ![image](https://github.com/directus/directus/assets/14810858/cbccb282-0563-42c1-82e6-5583e53330f2) | 

| Register Route | After Success |
|--------|--------|
| ![image](https://github.com/directus/directus/assets/14810858/a4b54235-fc42-4344-b825-68aaaa9fb0af) | ![image](https://github.com/directus/directus/assets/14810858/769d41ad-879b-49b3-934a-bed35a5332a5) | 

- Fixed UI inconsistency in login & reset-password

| Before | After |
|--------|--------|
| <video src="https://github.com/directus/directus/assets/14810858/13b888dc-dd3a-44ba-aeae-ed09aa949e60" autoplay loop muted></video> | <video autoplay loop muted src="https://github.com/directus/directus/assets/14810858/3d01ef66-041a-4998-ba7b-626eda9ad750"></video> | 


## Potential Risks / Drawbacks

- todo

## Review Notes / Questions

- Dont forget to run the migration :wink: 

---

Fixes #21981

DanielBiegler committed:
WIP: add new register dummy-route

DanielBiegler committed:
fix notice on register route

DanielBiegler committed:
WIP register form

DanielBiegler committed:
WIP: registering ui and controller for testing

DanielBiegler committed:
fix lint ordering problem

DanielBiegler committed:
wip: users service

DanielBiegler committed:
add migration, initial style for fields in settings

DanielBiegler committed:
redo how emails will be filtered

DanielBiegler committed:
WIP add filter in the register handler

DanielBiegler committed:
conditionally render register link depending on settings

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into fix-21981-public-registration

DanielBiegler committed:
WIP: add email validation

DanielBiegler committed:
wip add email sending

DanielBiegler committed:
make clicking the email link work

DanielBiegler committed:
rm console log

DanielBiegler committed:
update controller

DanielBiegler committed:
dont send emails for existing emails

DanielBiegler committed:
add translation

DanielBiegler committed:
only show register link when unauthenticated

DanielBiegler committed:
add different redirects

DanielBiegler committed:
only allow selecting non-admin roles

DanielBiegler committed:
redirect to users page

DanielBiegler committed:
update translation

DanielBiegler committed:
move logic from controller to usersservice

DanielBiegler committed:
rm remnant of logic from controller

DanielBiegler committed:
add stall time to registration

DanielBiegler committed:
update translation

DanielBiegler committed:
rm comments

DanielBiegler committed:
rm unused var

DanielBiegler commented on api/src/services/users.ts:
```diff
@@ -451,6 +458,111 @@ export class UsersService extends ItemsService {
+			email,
+			password,
+			role: publicRegistrationRole,
+			status: hasEmailValidation ? 'draft' : 'active', // TODO: Do we want to have a dedicated "unverified" status?
```
I'm thinking about renaming the `invited`-status to a more generic `pending`-status so that we can use the same status for invited and registered-but-not-yet-verified users but simply using the migration file to rename the status could lead to breaking existing logic in users projects for example somebody might have Flows that expect the users to be `invited`

Another approach could be to just replace the `Invited` UI label to `Pending` and keep the underlying `invited` string the same to stay backwards compatible. 

Mhh. Could also just add a new status to be both precise and backwards compatible. Like so:

![image](https://github.com/directus/directus/assets/14810858/d46af388-60ff-4b69-afa2-c318ea32d3db)

If we add the new status we can give it a more concrete name too like `Unverified` for example

paescuj replied:
I would have to take a closer look, but I think a dedicated (hidden) field for the email verification status could make sense. This could also be used, for example, if the user is already active and changes the email address, which would have to be verified again (but the user status should stay "active").

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into fix-21981-public-registration

DanielBiegler commented on app/src/routes/register/register.vue:
```diff
@@ -0,0 +1,139 @@
+
+		<register-form v-else :provider="provider" @was-successful="wasSuccessful = $event" />
+
+		<sso-links v-if="!authenticated && wasSuccessful == false" :providers="auth.providers" />
```
actually, do we want to show sso links on the register page? :thinking:

br41nslug replied:
> as those fall under the action of "login", right

Not necessarily if you have public registration enabled for SSO it doubles as both login and registration 🤔 
I wouldnt be against having the SSO buttons on the registration page if configured as such but feels like someone with more UX insight may have more opinions there.

DanielBiegler replied:
Agreed, removing them now

paescuj replied:
> Not necessarily if you have public registration enabled for SSO it doubles as both login and registration 🤔

Yeah, but you actually login to the provider with an existing account there (at least in most cases)

DanielBiegler committed:
add changeset

DanielBiegler commented on api/src/controllers/users.ts:
```diff
@@ -501,4 +501,42 @@ router.post(
 	respond,
 );
 
+const registerSchema = Joi.object<{ email: string; password: string }>({
+	email: Joi.string().email().required(),
+	password: Joi.string().required(),
+});
+
+router.post(
+	'/register',
```
It'd be nice if we would add a way to specify more user fields because I can see people using this service to build their own login pages with more fields :thinking: 

Allowing other user fields would need to be sanitized though we cant let them override role for example

paescuj replied:
You mean like require a phone number or similar? 🤔

DanielBiegler replied:
Yes for example a number or even simpler first & last name. Avatar could also be useful, stuff like that. It'd be nice to also allow populating user-provided fields for when users have manually extended the `directus_users` table so it'd have to be dynamic but I'm a little hesitant there because of security concerns. Mhh

paescuj replied:
Interesting ❤️ Well, actually it would "just" be another (public/shared) form where we could use existing validation logic, etc. 🤔

DanielBiegler committed:
update translation for success

paescuj commented on sdk/src/rest/commands/server/info.ts:
```diff
@@ -5,6 +5,7 @@ export type ServerInfoOutput = {
 	project: {
 		project_name: string;
 		default_language: string;
+		is_public_registration_enabled: boolean; // TODO Get opinion from @Brainslug
```
This should be sufficient IMO, the type already indicates that it's a toggle state:
```suggestion
		public_registration: boolean;
```

E.g. comparable to states in Collection metadata:
https://github.com/directus/directus/blob/d10385d30305adbe1b9abfa4bdefcff1836effca/packages/system-data/src/types.ts#L26-L33

DanielBiegler committed:
remove sso related stuff from registration

DanielBiegler committed:
also allow setting first and last name

hanneskuettner commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -1364,6 +1364,11 @@ fields:
     custom_aspect_ratios: Custom Aspect Ratios
     theme_light_overrides: Light Theme Customization
     theme_dark_overrides: Dark Theme Customization
+    public_registration: Public registration
+    is_public_registration_enabled: Public registration
```
These should probably both be title case 

```suggestion
    public_registration: Public Registration
    is_public_registration_enabled: Public Registration
```

DanielBiegler committed:
update error check

DanielBiegler commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,113 @@
+		password.value === null ||
+		passwordsMatch.value === false
+	) {
+		error.value = 'INVALID_PAYLOAD';
```
This made me check and turns out `@directus/errors` is currently not a dependency of the app?! :exploding_head:  Interesting. Good find, thanks.

DanielBiegler committed:
add @directus/errors to app

DanielBiegler committed:
replace error strings with enum

DanielBiegler committed:
rename to public_registration

DanielBiegler committed:
rename to public_registration_verify_email

DanielBiegler committed:
add notes to fields

DanielBiegler committed:
add types package to changeset

DanielBiegler committed:
dont stall if no work is being done

DanielBiegler committed:
allow null-role and resending of reg. email

licitdev commented on .changeset/dry-otters-attend.md:
```diff
@@ -0,0 +1,9 @@
+---
+'@directus/system-data': patch
```
Should this be a `minor` release to increment the `directus` version?

```suggestion
'directus': minor
'@directus/system-data': patch
```

paescuj commented on api/src/database/migrations/20240422A-public-registration.ts:
```diff
@@ -0,0 +1,31 @@
+import type { Knex } from 'knex';
+
+// To avoid typos
```
Bit of a nitpick, but since this migration is fairly small, I'd prefer to inline these values which would improve the readability IMO. That's also how we're doing it in the existing migrations.

DanielBiegler replied:
If it _strongly_ irks you or someone else, we can change it, but personally I dont mind the variables ¯\\\_(ツ)\_/¯

DanielBiegler replied:
Done

DanielBiegler committed:
add public registration env vars, rm RATE_LIMITER_GLOBAL_STORE

RATE_LIMITER_GLOBAL_STORE wasnt being used. Lets just stick to RATE_LIMITER_STORE for all rate limiters. TODO: also remove from docs!

DanielBiegler committed:
use ratelimiter for registration, use stall time env var

DanielBiegler committed:
add registration limiter docs, rm global store variable from docs

hanneskuettner commented on api/src/services/mail/templates/user-registration.liquid:
```diff
@@ -0,0 +1,36 @@
+<p style='padding-bottom: 30px'>
+  Thanks for registering with us at <i>{{ projectName }}</i>.
+  To complete your registration you need to verify your email address by opening the following verification-link.
+</p>
```
Should we add one of those standard disclaimers a la "If you don’t want to create an account, you can ignore this message."

DanielBiegler replied:
Yeah, sounds good, agreed! :+1:

DanielBiegler committed:
update changeset

hanneskuettner commented on api/src/services/users.ts:
```diff
@@ -451,6 +451,111 @@ export class UsersService extends ItemsService {
+			mailService
+				.send({
+					to: input.email,
+					subject: 'Verify your email address', // TODO: translate after theres support for internationalized emails
```
Should probably be title case? That's at least that's what all my emails I can see in my inbox have.

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into fix-21981-public-registration

hanneskuettner commented on packages/system-data/src/fields/settings.yaml:
```diff
@@ -200,6 +200,60 @@ fields:
+  - field: public_registration_role
+    interface: select-dropdown-m2o
+    note: $t:fields.directus_settings.public_registration_role_note
+    options:
```
Do we have the possibility to add validation rules here? That the user can not leave this `null` if the public registration is turned on? Don't know if that is possible in there definition files, but would be a nice to have.

hanneskuettner commented on packages/system-data/src/fields/settings.yaml:
```diff
@@ -200,6 +200,60 @@ fields:
+    display_options:
+      template: '{{ name }}'
+
+  - field: public_registration_verify_email
```
Personally I'd swap verify email and the email filter, since the email filter is more closely related to the user role (conceptionally) and the send verification email more related to a process of the registration as a whole.

DanielBiegler replied:
I prefer the toggles being on the left side while the more configurable fields live on the right. Since this is just an opinion thing I'd leave it up to @benhaynes / @rijkvanzanten - For quick reference here is how it currently looks like:

![image](https://github.com/directus/directus/assets/14810858/28e9efef-c9b6-424f-99c9-6f59398a9cf8)

hanneskuettner commented:
One whole new thought that came to me while reviewing this. Should there be an option to avoid creating any users until the user clicks the verification link? This would fully avoid having the possibility of filling a DB with bogus users.

Theoretically we can pass the full user payload through the signed token (which would not allow any tempering) and delay the account creation until the verification callback.

hanneskuettner commented on packages/system-data/src/fields/settings.yaml:
I think we talked about it, but I feel like the options for the registration should be disabled if the registration feature itself is disabled.

br41nslug replied:
In that same vein do we disable local registration when local auth is disabled? `AUTH_DISABLE_DEFAULT`

hanneskuettner replied:
Yes! And they can make changes without saving the page, can they not?

DanielBiegler replied:
> And they can make changes without saving the page, can they not?

Not sure if I understand you here? If I wanna just update the email filter I would then have to enable the registration, update the email filter, disable the registration again and then save. That doesnt sound useful to me

hanneskuettner commented on app/src/routes/register/register.vue:
```diff
@@ -0,0 +1,112 @@
+		</div>
+
+		<div v-if="wasSuccessful" class="after-success">
+			<div v-md="t('registration_successful_note')"></div>
```
Does it make sense to have two different notices depending on wether the verification email is actually enabled or not. UX wise it feels a bit weird and I could see users waiting for an email to arrive with the notice being worded like this.

hanneskuettner commented on packages/env/src/constants/defaults.ts:
```diff
@@ -29,7 +29,10 @@ export const DEFAULTS = {
 	RATE_LIMITER_GLOBAL_DURATION: 1,
-	RATE_LIMITER_GLOBAL_STORE: 'memory',
+
+	RATE_LIMITER_REGISTRATION_ENABLED: false,
```
Should this be enabled by default since we're basically handing the users a loaded gun here?

DanielBiegler replied:
Yeah thought the same thing, but we dont enable other rate limiters by default either. Both the normal and global one are disabled by default too. Mhh :thinking:

paescuj replied:
This is to be compared to the login rate limiter which cannot be turned off either 👍

DanielBiegler replied:
Not every environment cares about limiting this so I'd keep the ability to turn it off but maybe we indeed should enable it by default with conservative defaults

hanneskuettner commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -1364,6 +1364,16 @@ fields:
     theme_light_overrides: Light Theme Customization
     theme_dark_overrides: Dark Theme Customization
+    public_registration: User Registration
+    public_registration_note: Allows users to register via the [registration route](/admin/register).
```
Maybe we can use technical wording? 

```suggestion
    public_registration_note: Allows users to register via a [registration page](/admin/register).
```

paescuj replied:
Additionally, I'd probably leave out the link. Maybe something like this?
```suggestion
    public_registration_note: Allows users to self-register to the project.
```

DanielBiegler replied:
Linking is useful because not everyone will know where the registration will appear or how. Why would less information here be better than linking to the page?

paescuj replied:
The feature won't be limited to the register page of the Data Studio, but also for registering via API, so a more general note seems appropriate to me 👍

DanielBiegler replied:
> The feature won't be limited to the register page

Ahh true, good point! But I still like linking to it, I regard that as useful UX for admins. I used it quite often. mhh

paescuj replied:
Maybe a combination of both then?
```suggestion
    public_registration_note: Allows users to register via the [registration page](/admin/register) and API.
```

```suggestion
    public_registration_note: Allows users to self-register via the [registration page](/admin/register) and API.
```

DanielBiegler commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,114 @@
+			:disabled="isLoading"
+		/>
+		<v-input v-model="password" type="password" :placeholder="t('password')" :disabled="isLoading" />
+		<v-input v-model="passwordVerification" type="password" :placeholder="t('password')" :disabled="isLoading" />
```
The consensus was that the toggleable password input would be nice-to-have and preferred but so far I havent gotten around to it. As far as I'm aware, we would need a new component for that. :+1:

DanielBiegler committed:
add ignore-notice

Co-authored-by: Hannes Küttner <4376726+hanneskuettner@users.noreply.github.com>

paescuj commented on api/src/services/mail/templates/user-registration.liquid:
```diff
@@ -0,0 +1,36 @@
+<h1>Verify your email address</h1>
+
+<p style='padding-bottom: 30px'>
+  Thanks for registering with us at <i>{{ projectName }}</i>.
```
Or even
```suggestion
  You have registered at <i>{{ projectName }}</i>.
```

hanneskuettner commented on app/src/routes/register/register.vue:
IMHO this route should have a `beforeEnter` guard that redirects logged in users to either `/login` or directly to their last page. It feels unfinished if a user can view the register page as a logged in user.

licitdev replied:
+1. 👍 This goes with Pascal's comment to remove the link to this page. https://github.com/directus/directus/pull/22125/files#r1591511293

DanielBiegler replied:
If theres no technical reason for forbidding logged in users reaching that page, I dont see the value. Under normal usage for normal users, you cant even see the link to that route on the log-in page because of the "you are currently logged in as"-component. So non-admins would have to manually type in that url or log out. And at that point just let people go there. Admins will see the link in the project settings but normal users dont. Without a technical reason this seems unimportant to me. If we decide that we need to forbid api usage of that route while someone is logged in, then yes, we should also redirect them but thats currently not the case.

licitdev replied:
I rest my case. 👍  Just noticed that the `/reset-password` route works the same way as `/register`.

paescuj replied:
Generally, I agree that there is no compelling reason to hide this page for logged-in users.
However, what's a bit confusing at the moment, is that when you register while already logged-in, the success message appears, with the "sign in" button, even though you are (obviously) already signed-in as current user - maybe a "switch user" button or similar would make that clearer...

On a similar note, the register page is currently available, even when public registration is disabled - but I guess that's fine too.

DanielBiegler commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -1364,6 +1364,16 @@ fields:
+      This role is assigned to users who register using this method. Does not affect already registered users.
+    public_registration_email_filter: Email Address Filter
+    public_registration_email_filter_note: Only email addresses matching this filter can register.
+    public_registration_verify_email: Verify Email
```
I had send verification email beforehand and Ben mentioned in the show&tell we can shorten it to verify email. Leaving it for now. We can change it later if necessary :+1:

DanielBiegler commented on api/src/services/users.ts:
```diff
@@ -451,6 +458,118 @@ export class UsersService extends ItemsService {
+		if (hasEmailValidation) {
+			const mailService = new MailService(serviceOptions);
+			const payload = { email: input.email, scope: 'pending-registration' };
+			const token = jwt.sign(payload, env['SECRET'] as string, { expiresIn: '7d', issuer: 'directus' });
```
Mhh initially I agreed because I thought we can reuse it for more tokens but now I looked up our other TTLs and theres no further overlap, just the invite and registration. Furthermore it sounds like this could be a nice configuration env variable even! For example we also have `SESSION_COOKIE_TTL`, `ACCESS_TOKEN_TTL` so why not add something like `EMAIL_VERIFICATION_TOKEN_TTL`, I can already see other people wanting to override this to shorter time frames. I have not thought about this initially, thanks! :heart:  :+1:

DanielBiegler committed:
use and document new `EMAIL_VERIFICATION_TOKEN_TTL`, also doc `REGISTER_STALL_TIME`

DanielBiegler committed:
change variable name

Co-authored-by: ian <licitdev@gmail.com>

DanielBiegler committed:
apply variable rename to usage

DanielBiegler committed:
change backticks to single quote

Co-authored-by: ian <licitdev@gmail.com>

DanielBiegler committed:
inline variables

DanielBiegler committed:
Merge remote-tracking branch 'refs/remotes/origin/fix-21981-public-registration' into fix-21981-public-registration

licitdev commented on api/src/services/users.ts:
```diff
@@ -452,6 +452,111 @@ export class UsersService extends ItemsService {
+			// To avoid giving attackers infos about registered emails we dont fail for violated unique constraints
+			await stall(STALL_TIME, timeStart);
+			return;
+		}
```
Yes, the user can still click on the verification link to activate their account if the token is still valid. It's an edge case for users in the pending verification `draft` status with expired tokens where the email verification requirement was disabled after their registration.

The logical check in the preceding block ensures that `user.status === 'draft'`. 
https://github.com/directus/directus/blob/854dd1c1d7b67d560a5a6b19a8fb824300efdc9e/api/src/services/users.ts#L503-L507

I can't think of a scenario right now where it backfires. 🤔

hanneskuettner replied:
Would an edge case be, similar to the one you mentioned below, where this could result in a 'draft' admin account being activated through this flow?

hanneskuettner replied:
Which already is a problem of the current version, tbh

licitdev replied:
Yes, they are kind of related as both involve existing `draft` accounts. However, with the introduction of a new status, this edge case remains.

The logical block linked above could be updated with the new `status`, but the code suggestion remains.

hanneskuettner replied:
So the "attack" would be that a new admin user has been set up, but is still intended to by in the `draft` status until another admin actives the account. 
That user can, if public registration is enabled and email verification disabled self activate their account, reset their password, as they now have an active account and then log in to the system, even though that user was never admitted as an active admin.
That seems unlikely, but still a valid edge case that should not be possible IMO.

licitdev reviewed (commented):
Will the `graphql` and `sdk` commands for public user registration be included in this PR?

DanielBiegler committed:
add fields to server info, update types

- The other ratelimiters also expose points and duration, done
- Add `public_registration_verify_email` so that we can render different success messages

br41nslug committed:
tiny wording tweak of registration mail

DanielBiegler committed:
add new user status 'unverified' and check for it

DanielBiegler committed:
Merge remote-tracking branch 'refs/remotes/origin/fix-21981-public-registration' into fix-21981-public-registration

br41nslug commented on api/src/middleware/rate-limiter-registration.ts:
We can abstract this in a later optimization if need be.

DanielBiegler committed:
add unverified status translation

DanielBiegler committed:
decouple email verification and validation

DanielBiegler commented on api/src/services/users.ts:
```diff
@@ -452,6 +452,115 @@ export class UsersService extends ItemsService {
+			email: input.email,
+			password: input.password,
+			role: publicRegistrationRole,
+			status: hasEmailVerification ? 'draft' : 'active', // TODO: Do we want to have a dedicated "unverified" status?
```
Done :)

DanielBiegler committed:
enable register rate limiter by default and up its config

hanneskuettner commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,114 @@
+			:placeholder="t('email')"
+			:disabled="isLoading"
+		/>
+		<v-input v-model="password" type="password" :placeholder="t('password')" :disabled="isLoading" />
```
Nitpick: Add `autocomplete="new-password"` to play nice with browsers and password managers.

```suggestion
		
		<v-input
			v-model="password"
			type="password"
			autocomplete="new-password"
			:placeholder="t('password')"
			:disabled="isLoading"
		/>
```

br41nslug replied:
nice one 😄

hanneskuettner commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,114 @@
+			:disabled="isLoading"
+		/>
+		<v-input v-model="password" type="password" :placeholder="t('password')" :disabled="isLoading" />
+		<v-input v-model="passwordVerification" type="password" :placeholder="t('password')" :disabled="isLoading" />
```
Same as above

```suggestion
		<v-input
			v-model="passwordVerification"
			type="password"
			autocomplete="new-password"
			:placeholder="t('password')"
			:disabled="isLoading"
		/>
```

br41nslug committed:
add autocomplete=new-password on the registration form

br41nslug committed:
added sdk functions

DanielBiegler committed:
add gql query for new fields

DanielBiegler committed:
Merge remote-tracking branch 'refs/remotes/origin/fix-21981-public-registration' into fix-21981-public-registration

br41nslug committed:
added register api reference

br41nslug committed:
Merge branch 'fix-21981-public-registration' of https://github.com/directus/directus into fix-21981-public-registration

br41nslug committed:
updated verify sdk function name

br41nslug committed:
added reference block for email verify endpoint

paescuj commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -1364,6 +1364,16 @@ fields:
+    public_registration_email_filter_note: Only email addresses matching this filter can register.
+    public_registration_verify_email: Verify Email
+    public_registration_verify_email_note:
+      Sends an email to the registering user asking them to click on a verification link before they can sign in.
```
```suggestion
      Send an email to newly signed-up users to verify their email address before they can sign in.
```

br41nslug committed:
updated reference examples

paescuj commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -1631,6 +1642,14 @@ select_field_type: Select a field type
+registration_successful_note: You may now proceed to the sign in page.
+registration_successful_check_email_note: |
+  Please note that before you can sign in, you'll need to verify your email address by clicking on the verification link sent to you.
+dont_have_an_account: Don't have an account?
+already_have_an_account: Already have an account?
```
Translation keys should be somewhat neutral (and existing ones generally are)
```suggestion
no_account: Don't have an account?
already_an_account: Already have an account?
```

DanielBiegler committed:
WIP: add gql resolvers

paescuj commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,126 @@
+const passwordsMatch = computed(() => password.value === passwordVerification.value);
+
+const emit = defineEmits<{
+	wasSuccessful: [boolean];
```
AFAICS, we only seem to care about the event itself
```suggestion
	success: [];
```

paescuj commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,126 @@
+}>();
+
+const errorFormatted = computed(() => {
+	// Show "Wrong username or password" for wrongly formatted emails as well
+	if (error.value === ErrorCode.InvalidPayload) {
+		return translateAPIError(ErrorCode.InvalidCredentials);
+	}
```
This message makes sense for login, but I think for registering it should be: Invalid username or password.

paescuj commented on app/src/routes/register/register-form.vue:
```diff
@@ -0,0 +1,126 @@
+
+		await api.post('/users/register', credentials);
+
+		emit('wasSuccessful', true);
+	} catch (err: any) {
+		error.value = err.errors?.[0]?.extensions?.code || err;
+		emit('wasSuccessful', false);
```
As mentioned above
```suggestion
		emit('success');
	} catch (err: any) {
		error.value = err.errors?.[0]?.extensions?.code || err;
```

paescuj commented on app/src/routes/register/register.vue:
```diff
@@ -0,0 +1,119 @@
+			<v-button large to="/login">{{ t('sign_in') }}</v-button>
+		</div>
+
+		<register-form v-else @was-successful="wasSuccessful = $event" />
```
```suggestion
		<register-form v-else @success="wasSuccessful = true" />
```

DanielBiegler committed:
add ratelimiter to mutation

paescuj commented on api/src/controllers/users.ts:
```diff
@@ -501,4 +502,45 @@ router.post(
+const verifyRegistrationSchema = Joi.string();
+
+router.get(
+	'/register/verify-email',
```
Wondering if this should just be `/register/verify`, similar to other single-worded routes like `/invite/accept`, etc. and also to be in-line with the service method name, SDK command, ...

paescuj reviewed (commented):
- What terminology have we agreed on? In the Data Studio "User Registration" is used, while the API configs use "public_registration". This should be aligned before merging.
- Docs are missing for the new Project Settings [^1], not sure if we still want to maintain this page, though.

[^1]: https://docs.directus.io/user-guide/settings/project-settings.html

br41nslug reviewed (approved):
🚀

DanielBiegler commented on api/src/services/server.ts:
```diff
@@ -78,6 +80,15 @@ export class ServerService {
 				info['rateLimitGlobal'] = false;
 			}
 
+			if (env['RATE_LIMITER_REGISTRATION_ENABLED']) {
+				info['rateLimitRegistration'] = {
+					points: env['RATE_LIMITER_REGISTRATION_POINTS'],
+					duration: env['RATE_LIMITER_REGISTRATION_DURATION'],
+				};
+			} else {
+				info['rateLimitRegistration'] = false;
+			}
```
Good point, on it

DanielBiegler committed:
remove ratelimiter registration point+duration info

DanielBiegler committed:
rm points and duration from gql

br41nslug committed:
Update docs/reference/system/users.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
